src/Microsoft.AspNetCore.Http/HttpContextAccessor.cs

```diff
@@ -7,17 +7,32 @@ namespace Microsoft.AspNetCore.Http
 {
     public class HttpContextAccessor : IHttpContextAccessor
     {
-        private static AsyncLocal<HttpContext> _httpContextCurrent = new AsyncLocal<HttpContext>();
+        private static AsyncLocal<(string traceIdentifier, HttpContext context)> _httpContextCurrent = new AsyncLocal<(string traceIdentifier, HttpContext context)>();
 
         public HttpContext HttpContext
         {
             get
             {
-                return _httpContextCurrent.Value;
+                var value = _httpContextCurrent.Value;
+                // Only return the context if the stored request id matches the stored trace identifier
+                return value.traceIdentifier == value.context?.TraceIdentifier ? value.context : null;
             }
             set
             {
-                _httpContextCurrent.Value = value;
+                var existing = _httpContextCurrent.Value;
+
+                if (value != null)
+                {
+                    // Store the request id and the HttpContext
+                    _httpContextCurrent.Value = (value.TraceIdentifier, value);
+                }
+                else
+                {
+                    // Setting the context to null means the request is over
+                    existing.context.TraceIdentifier = null;
```

Comment:
Is set to `null` in `HttpProtocol.Reset()`?

Reply:
WebListener doesn't do this today so I'm not sure I want to remove it yet. It means all server implementations need to mutate the `HttpContext` on Dispose.

Reply:
That's better than having an arbitrary field mutated the accessor. Any way to give it its own field? A feature?

Reply:
I originally had that but it will make the performance worse (as it modifies the feature. We already have a feature that I'm piggy backing on. I would need to make a new feature and implement it in all the server feature collections then clear it here. It's exactly the same as this.

So the question is, what's the downside?